File: persipubsub/queue.py
#!/usr/bin/env python3
"""Store messages in a local LMDB."""
import datetime
import enum
import pathlib
import uuid
from typing import Any, Dict, List, Optional, Union

import lmdb

import persipubsub

MSG_TIMEOUT_SECS = 500  # type: int
MAX_MSGS_NUM = 1024 * 64
HWM_LMDB_SIZE_BYTES = 30 * 1024**3  # type: int


class Strategy(enum.Enum):
    """Hold possible strategies."""

    prune_first = 0
    prune_last = 1


def _parse_strategy(strategy: str) -> Strategy:
    """
    Parse overflow strategy.

    :param strategy: Strategy stored in LMDB
    :return: set overflow strategy
    """
    if Strategy.prune_first.name == strategy:
        return Strategy.prune_first

    if Strategy.prune_last.name == strategy:
        return Strategy.prune_last

    raise ValueError("Unknown strategy: '{}' choosen.".format(strategy))


class HighWaterMark:
    """Hold high water mark limits."""

    def __init__(self,
                 msg_timeout_secs: Optional[int] = None,
                 max_msgs_num: Optional[int] = None,
                 hwm_lmdb_size_bytes: Optional[int] = None) -> None:
        """
        Initialize.

        :param msg_timeout_secs: time after which msg is classified as dangling
        msg (secs)
        :param max_msgs_num: maximal amount of msg
        :param hwm_lmdb_size_bytes: high water mark for total size of LMDB
        (bytes)
        """
        self.msg_timeout_secs = MSG_TIMEOUT_SECS \
            if msg_timeout_secs is None else msg_timeout_secs  # type: int
        self.max_msgs_num = MAX_MSGS_NUM \
            if max_msgs_num is None else max_msgs_num  # type: int
        self.hwm_lmdb_size_bytes = HWM_LMDB_SIZE_BYTES \
            if hwm_lmdb_size_bytes is None else hwm_lmdb_size_bytes  # type: int


def _initialize_environment(
        queue_dir: pathlib.Path,
        max_reader_num: int = 1024,
        max_db_num: int = 1024,
        max_db_size_bytes: int = 32 * 1024**3) -> lmdb.Environment:
    """
    Initialize the queue; the queue directory is assumed to exist.

    :param queue_dir: where the queue is stored
    :param max_reader_num: maximal number of readers
    :param max_db_num: maximal number of databases
    :param max_db_size_bytes: maximal size of database (bytes)
    :return: Load or if needed create LMDB from directory
    """
    if not queue_dir.exists():
        raise RuntimeError(
            "The queue directory does not exist: {}".format(queue_dir))

    # max_spare_txn: Read-only transactions to cache after becoming unused.
    # Caching transactions avoids two allocations, one lock and linear scan of
    # the shared environment per invocation of begin(), Transaction, get(),
    # gets(), or cursor(). Should match the process’s maximum expected
    # concurrent transactions (e.g. thread count).
    # Setting max_spare_txn equals 0 doesn't allow any caching for read-only
    # transactions to avoid the error:
    # MDB_BAD_RSLOT: Invalid reuse of reader locktable slot.
    # Error only occures when one publisher/subscriber is used in multiple
    # processes. So max_spare_txn equal 1 is fine as we expect a publisher and
    # subscriber running on the same process.
    env = lmdb.open(
        path=queue_dir.as_posix(),
        map_size=max_db_size_bytes,
        subdir=True,
        max_readers=max_reader_num,
        max_dbs=max_db_num,
        max_spare_txns=1)
    return env


def _prune_dangling_messages_for(queue: '_Queue',
                                 subscriber_ids: List[str]) -> None:
    """
    Prune all dangling messages for subscribers of a queue from LMDB.

    :param queue: of which dangling messages should be pruned
    :param subscriber_ids: subscribers of which dangling msgs should be pruned
    """
    assert isinstance(queue.env, lmdb.Environment)
    with queue.env.begin(write=True) as txn:
        pending_db = queue.env.open_db(
            key=persipubsub.PENDING_DB, txn=txn, create=False)
        meta_db = queue.env.open_db(
            key=persipubsub.META_DB, txn=txn, create=False)
        data_db = queue.env.open_db(
            key=persipubsub.DATA_DB, txn=txn, create=False)

    # Definition of dangling messages:
    #   - having no pending subscribers
    #   - exists longer than timeout allows
    msgs_to_delete = set()
    with queue.env.begin(db=pending_db) as txn:
        cursor = txn.cursor()
        cursor.first()
        for key, pending_subscribers_num in cursor:
            if persipubsub.bytes_to_int(pending_subscribers_num) == 0:
                msgs_to_delete.add(key)

    # subscriber might still await these messages after the timeout.
    # These messages needs also to be removed from all subscribers.
    msgs_to_delete_timeout = set()
    assert isinstance(queue.hwm, HighWaterMark)
    with queue.env.begin(db=meta_db) as txn:
        cursor = txn.cursor()

        timestamp_now = datetime.datetime.utcnow().timestamp()
        for key, timestamp in cursor:
            if int(timestamp_now) - persipubsub.bytes_to_int(timestamp) \
                    > queue.hwm.msg_timeout_secs:
                msgs_to_delete_timeout.add(key)

    msgs_to_delete = msgs_to_delete.union(msgs_to_delete_timeout)
    with queue.env.begin(write=True) as txn:
        for key in msgs_to_delete:
            txn.delete(key=key, db=pending_db)
            txn.delete(key=key, db=meta_db)
            txn.delete(key=key, db=data_db)

    with queue.env.begin(write=True) as txn:
        for sub_id in subscriber_ids:
            sub_db = queue.env.open_db(
                key=persipubsub.str_to_bytes(sub_id), txn=txn, create=False)
            for key in msgs_to_delete_timeout:
                txn.delete(key=key, db=sub_db)


class _Queue:
    """
    Represent a message queue.

    The messages come from many publishers and many subscribers.
    They are stored persistently in an LMDB database.

    :ivar path: to the queue
    :vartype config_pth: pathlib.Path
    :ivar env: LMDB environment on disk
    :vartype env: lmdb.Environment
    :ivar hwm: High water mark limit for queue
    :vartype hwm: HighWaterMark
    :ivar strategy:
        Strategy which will be used to remove messages when high water mark is
        reached.
    :vartype strategy: Strategy
    :ivar subscriber_ids: all subscribers of the queue
    :vartype sub_list: List[str]
    """

    def __init__(self) -> None:
        """Initialize class object."""
        self.path = None  # type: Optional[pathlib.Path]
        self.env = None  # type: Optional[lmdb.Environment]
        self.hwm = None  # type: Optional[HighWaterMark]
        self.strategy = None  # type: Optional[Strategy]
        self.subscriber_ids = None  # type: Optional[List[str]]

    def init(self,
             path: Union[pathlib.Path, str],
             env: Optional[lmdb.Environment] = None) -> None:
        """
        Initialize the queue.

        :param path: where the queue is stored
        :param env: open LMDB environment
        """
        # pylint: disable=too-many-arguments
        # pylint: disable=too-many-locals
        self.path = path if isinstance(path, pathlib.Path) \
            else pathlib.Path(path)

        if isinstance(env, lmdb.Environment):
            self.env = env
        else:
            self.env = _initialize_environment(
                queue_dir=self.path,
                max_reader_num=persipubsub.MAX_READER_NUM,
                max_db_num=persipubsub.MAX_DB_NUM,
                max_db_size_bytes=persipubsub.MAX_DB_SIZE_BYTES)

        with self.env.begin(write=True) as txn:
            _ = self.env.open_db(key=persipubsub.DATA_DB, txn=txn, create=True)
            _ = self.env.open_db(
                key=persipubsub.PENDING_DB, txn=txn, create=True)
            _ = self.env.open_db(key=persipubsub.META_DB, txn=txn, create=True)
            _ = self.env.open_db(key=persipubsub.QUEUE_DB, txn=txn, create=True)

        msg_timeout_secs_bytes = persipubsub.lookup_queue_data(
            key=persipubsub.MSG_TIMEOUT_SECS_KEY, env=self.env)
        msg_timeout_secs = msg_timeout_secs_bytes \
            if msg_timeout_secs_bytes is None else persipubsub.bytes_to_int(
            msg_timeout_secs_bytes)  # type: Optional[int]

        max_msgs_num_bytes = persipubsub.lookup_queue_data(
            key=persipubsub.MAX_MSGS_NUM_KEY, env=self.env)
        max_msgs_num = max_msgs_num_bytes \
            if max_msgs_num_bytes is None else persipubsub.bytes_to_int(
            max_msgs_num_bytes)  # type: Optional[int]

        hwm_lmdb_size_bytes = persipubsub.lookup_queue_data(
            key=persipubsub.HWM_DB_SIZE_BYTES_KEY, env=self.env)
        hwm_lmdb_size = hwm_lmdb_size_bytes \
            if hwm_lmdb_size_bytes is None else persipubsub.bytes_to_int(
            hwm_lmdb_size_bytes)  # type: Optional[int]

        self.hwm = HighWaterMark(
            msg_timeout_secs=msg_timeout_secs,
            max_msgs_num=max_msgs_num,
            hwm_lmdb_size_bytes=hwm_lmdb_size)

        strategy = persipubsub.lookup_queue_data(
            key=persipubsub.STRATEGY_KEY, env=self.env)

        strategy_decoded = "" if strategy is None else persipubsub.bytes_to_str(
            encoded_str=strategy)

        self.strategy = _parse_strategy(strategy=strategy_decoded)

        subscriber_list = persipubsub.lookup_queue_data(
            key=persipubsub.SUBSCRIBER_IDS_KEY, env=self.env)

        if subscriber_list is None:
            self.subscriber_ids = []
        else:
            self.subscriber_ids = persipubsub.bytes_to_str(
                encoded_str=subscriber_list).split(' ')

    def __enter__(self) -> '_Queue':
        """Enter the context and give the queue prepared in the constructor."""
        return self

    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        """Leave the context."""

    def put(self, msg: bytes) -> None:
        """
        Put message to LMDB in one transaction.

        :param msg: message in bytes
        :return:
        """
        # every publisher always prunes queue before sending a message.
        self.vacuum()
        msg_id = str(datetime.datetime.utcnow().timestamp()) + str(uuid.uuid4())
        assert isinstance(self.env, lmdb.Environment)
        assert isinstance(self.subscriber_ids, List)
        with self.env.begin(write=True) as txn:
            pending_db = self.env.open_db(
                key=persipubsub.PENDING_DB, txn=txn, create=False)
            txn.put(
                key=persipubsub.str_to_bytes(msg_id),
                value=persipubsub.int_to_bytes(len(self.subscriber_ids)),
                db=pending_db)

            meta_db = self.env.open_db(
                key=persipubsub.META_DB, txn=txn, create=False)
            txn.put(
                key=persipubsub.str_to_bytes(msg_id),
                value=persipubsub.int_to_bytes(
                    int(datetime.datetime.utcnow().timestamp())),
                db=meta_db)

            data_db = self.env.open_db(
                key=persipubsub.DATA_DB, txn=txn, create=False)
            txn.put(key=persipubsub.str_to_bytes(msg_id), value=msg, db=data_db)

            for sub in self.subscriber_ids:
                sub_db = self.env.open_db(
                    key=persipubsub.str_to_bytes(sub), txn=txn, create=False)
                txn.put(key=persipubsub.str_to_bytes(msg_id), db=sub_db)

    def put_many_flush_once(self, msgs: List[bytes]) -> None:
        """
        Put multiple message to LMDB in one transaction.

        :param msgs: messages in bytes
        :return:
        """
        # every publisher always prunes queue before sending a message.
        self.vacuum()
        assert isinstance(self.env, lmdb.Environment)
        assert isinstance(self.subscriber_ids, List)
        with self.env.begin(write=True) as txn:

            pending_db = self.env.open_db(
                key=persipubsub.PENDING_DB, txn=txn, create=False)

            meta_db = self.env.open_db(
                key=persipubsub.META_DB, txn=txn, create=False)

            data_db = self.env.open_db(
                key=persipubsub.DATA_DB, txn=txn, create=False)

            sub_dbs = set()
            for sub in self.subscriber_ids:
                sub_dbs.add(
                    self.env.open_db(
                        key=persipubsub.str_to_bytes(sub),
                        txn=txn,
                        create=False))

            for msg in msgs:
                msg_id = str(datetime.datetime.utcnow().timestamp()) + str(
                    uuid.uuid4())

                txn.put(
                    key=persipubsub.str_to_bytes(msg_id),
                    value=persipubsub.int_to_bytes(len(self.subscriber_ids)),
                    db=pending_db)

                txn.put(
                    key=persipubsub.str_to_bytes(msg_id),
                    value=persipubsub.int_to_bytes(
                        int(datetime.datetime.utcnow().timestamp())),
                    db=meta_db)

                txn.put(
                    key=persipubsub.str_to_bytes(msg_id), value=msg, db=data_db)

                for sub_db in sub_dbs:
                    txn.put(key=persipubsub.str_to_bytes(msg_id), db=sub_db)

    def front(self, identifier: str) -> Optional[bytes]:
        """
        Peek at next message in LMDB.

        Load from LMDB into memory and process msg afterwards.

        :param identifier: Subscriber ID
        :return:
        """
        assert isinstance(self.env, lmdb.Environment)
        with self.env.begin(write=False) as txn:
            sub_db = self.env.open_db(
                key=persipubsub.str_to_bytes(identifier), txn=txn, create=False)
            data_db = self.env.open_db(
                key=persipubsub.DATA_DB, txn=txn, create=False)

            cursor = txn.cursor(db=sub_db)
            # check if database is not empty
            if cursor.first():
                key = cursor.key()
                msg = txn.get(key=key, db=data_db)
            else:
                msg = None

        return msg  # type: ignore

    def pop(self, identifier: str) -> None:
        """
        Remove msg from the subscriber's queue and reduce pending subscribers.

        :param identifier: Subscriber ID
        :return:
        """
        assert isinstance(self.env, lmdb.Environment)
        with self.env.begin(write=True) as txn:
            sub_db = self.env.open_db(
                key=persipubsub.str_to_bytes(identifier), txn=txn, create=False)
            pending_db = self.env.open_db(
                key=persipubsub.PENDING_DB, txn=txn, create=False)

            cursor = txn.cursor(db=sub_db)
            # check if database is not empty
            if cursor.first():
                key = cursor.key()
                cursor.pop(key=key)
                pending_value = txn.get(key=key, db=pending_db)
                pending_num = persipubsub.bytes_to_int(pending_value)
                decreased_pending_num = pending_num - 1
                assert decreased_pending_num >= 0
                txn.put(
                    key=key,
                    value=persipubsub.int_to_bytes(decreased_pending_num),
                    db=pending_db)
            else:
                raise RuntimeError("No message to pop")

    def prune_dangling_messages(self) -> None:
        """
        Prune dangling messages in the queue.

        Definition of dangling messages:
        - having no pending subscribers
        - exists longer than timeout allows
        :return:
        """
        assert isinstance(self.subscriber_ids, List)
        _prune_dangling_messages_for(
            queue=self, subscriber_ids=self.subscriber_ids)

    def check_current_lmdb_size(self) -> int:
        """
        Check current LMDB size in bytes.

        Check size of data database by approximating size with multiplying page
        size with number of pages.

        :return: data database size in bytes
        """
        lmdb_size_bytes = 0
        assert isinstance(self.env, lmdb.Environment)
        with self.env.begin(write=False) as txn:
            data_db = self.env.open_db(
                key=persipubsub.DATA_DB, txn=txn, create=False)
            data_stat = txn.stat(db=data_db)
            lmdb_size_bytes += data_stat['psize'] * (
                data_stat['branch_pages'] + data_stat['leaf_pages'] +
                data_stat['overflow_pages'])

        return lmdb_size_bytes

    def count_msgs(self) -> int:
        """
        Count number of messages in database.

        Count number of messages stored in named database 'meta_db'.

        :return: number of messages in database
        """
        assert isinstance(self.env, lmdb.Environment)
        with self.env.begin(write=False) as txn:
            meta_db = self.env.open_db(
                key=persipubsub.META_DB, txn=txn, create=False)
            meta_stat = txn.stat(db=meta_db)  # type: Dict[str, int]

        return meta_stat['entries']

    def vacuum(self) -> None:
        """
        Clean database when needed.

        :return:
        """
        assert isinstance(self.hwm, HighWaterMark)
        self.prune_dangling_messages()
        msgs_num = self.count_msgs()
        if msgs_num >= self.hwm.max_msgs_num:
            self.prune_messages()
        lmdb_size_bytes = self.check_current_lmdb_size()
        if lmdb_size_bytes >= self.hwm.hwm_lmdb_size_bytes:
            self.prune_messages()

    def prune_messages(self) -> None:
        """
        Prune one half of the messages stored.

        Depending on the strategy the first or the last will be deleted.

        :return:
        """
        # pylint: disable=too-many-locals
        messages_to_delete = set()
        assert isinstance(self.env, lmdb.Environment)
        with self.env.begin(write=False) as txn:
            meta_db = self.env.open_db(
                key=persipubsub.META_DB, txn=txn, create=False)
            meta_stat = txn.stat(db=meta_db)
            entries = meta_stat['entries']

            cursor = txn.cursor(db=meta_db)
            if self.strategy == Strategy.prune_first:

                cursor.first()
                for index, key in enumerate(
                        cursor.iternext(keys=True, values=False)):
                    messages_to_delete.add(key)
                    if index >= int(entries / 2):
                        break

            elif self.strategy == Strategy.prune_last:
                cursor.last()
                for index, key in enumerate(
                        cursor.iterprev(keys=True, values=False)):
                    messages_to_delete.add(key)
                    if index >= int(entries / 2):
                        break
            else:
                raise RuntimeError("Pruning strategy not set.")

        assert isinstance(self.subscriber_ids, List)
        with self.env.begin(write=True) as txn:
            pending_db = self.env.open_db(
                key=persipubsub.PENDING_DB, txn=txn, create=False)
            meta_db = self.env.open_db(
                key=persipubsub.META_DB, txn=txn, create=False)
            data_db = self.env.open_db(
                key=persipubsub.DATA_DB, txn=txn, create=False)

            dbs = [pending_db, meta_db, data_db]

            for sub in self.subscriber_ids:
                sub_db = self.env.open_db(
                    key=persipubsub.str_to_bytes(sub), txn=txn, create=False)
                dbs.append(sub_db)

            for key in messages_to_delete:
                for db in dbs:  # pylint: disable=invalid-name
                    txn.delete(key=key, db=db)

File: cpp/src/queue.cpp
// Copyright (c) 2019 Parquery AG. All rights reserved.
// Created by Selim Naji (selim.naji@parquery.com/marko@parquery.com)
// on 15.02.2019

#include "queue.h"

//
// Created by selim on 01.02.19.
//

#include "library.h"
#include <lmdb++.h>
#include <boost/filesystem/path.hpp>
#include <boost/filesystem/operations.hpp>
#include <boost/numeric/conversion/cast.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/uuid/uuid.hpp>
#include <boost/uuid/uuid_generators.hpp>
#include <boost/uuid/uuid_io.hpp>
#include <vector>
#include <map>
#include <time.h>
#include <set>

namespace fs = boost::filesystem;

std::map<std::string, persipubsub::queue::Strategy> strategyMap;

constexpr char const_prune_first[] = "prune_first";
constexpr char const_prune_last[] = "prune_last";

persipubsub::queue::Strategy persipubsub::queue::parse_strategy(const std::string &strategy) {
    strategyMap[const_prune_first] = persipubsub::queue::Strategy::prune_first;
    strategyMap[const_prune_last] = persipubsub::queue::Strategy::prune_last;

    auto result = strategyMap.find(strategy);
    if (result == strategyMap.end())
        throw std::runtime_error("Unknown strategy: " + strategy);
    return result->second;


}


lmdb::env persipubsub::queue::initialize_environment(const fs::path& queue_dir,
                                 const unsigned int max_reader_num,
                                 const unsigned int max_db_num,
                                 const unsigned long max_db_size_bytes) {

    if (not fs::exists(queue_dir))
        throw std::runtime_error("The queue directory does not exist: " +
                                 queue_dir.filename().string());

    auto env = lmdb::env::create();
    env.set_mapsize(max_db_size_bytes);
    env.set_max_dbs(max_db_num);
    env.set_max_readers(max_reader_num);
    env.open(queue_dir.filename().string().c_str(), 0, 0664);
    return env;

}


void persipubsub::queue::prune_dangling_messages_for(const persipubsub::queue::Queue& queue,
                                 const std::vector<std::string> &subscriber_ids){

    /*todo check mistake:
     *  terminate called after throwing an instance of 'lmdb::runtime_error'
        what():  mdb_txn_commit: Invalid argument
        unknown location(0): fatal error in "test_put_to_single_subscriber": signal: SIGABRT (application abort requested)
        /home/selim/workspace/pqry/persipubsub/cpp/test/test_queue.cpp(99): last checkpoint
        terminate called recursively
     */

    auto wtxn = persipubsub::WriteTransaction(queue.env_).wtxn_;
    //auto wtxn = lmdb::txn::begin(queue.env_);
    auto pending_dbi = lmdb::dbi::open(wtxn, persipubsub::PENDING_DB);
    auto meta_dbi = lmdb::dbi::open(wtxn, persipubsub::META_DB);
    auto data_dbi = lmdb::dbi::open(wtxn, persipubsub::DATA_DB);

    // Definition of dangling messages:
    //   - having no pending subscribers
    //   - exists longer than timeout allows
    std::set<lmdb::val> msgs_to_delete;

    auto pending_cursor = lmdb::cursor::open(wtxn, pending_dbi);

    lmdb::val pending_key, pending_subscribers_num;
    while (pending_cursor.get(pending_key, pending_subscribers_num, MDB_NEXT)) {
        if (std::stoi(pending_subscribers_num.data()) == 0)
            msgs_to_delete.insert(std::move(pending_key));
    }
    pending_cursor.close();

    time_t timer;
    time(&timer);
    const auto timestamp_now = boost::numeric_cast<int>(timer);

    std::set<lmdb::val> msgs_to_delete_timeout;

    auto meta_cursor = lmdb::cursor::open(wtxn, pending_dbi);

    lmdb::val meta_key, timestamp;
    while (meta_cursor.get(meta_key, timestamp, MDB_NEXT)) {
        if (timestamp_now - std::stoi(timestamp.data()) > queue.hwm_.msg_timeout_secs_)
            msgs_to_delete_timeout.insert(std::move(meta_key));
    }
    meta_cursor.close();

    for (const auto& timeout_msg: msgs_to_delete_timeout){
        lmdb::val new_timeout_msg((timeout_msg).data());
        msgs_to_delete.insert(std::move(new_timeout_msg));
    }

    for (const auto& delete_it: msgs_to_delete){
        pending_dbi.del(wtxn, delete_it);
        meta_dbi.del(wtxn, delete_it);
        data_dbi.del(wtxn, delete_it);
    }

    for (const auto& sub_it: queue.subscriber_ids_){
        auto sub_dbi = lmdb::dbi::open(wtxn, (sub_it).c_str());
        for (const auto& key_it: msgs_to_delete_timeout){
            sub_dbi.del(wtxn, key_it);
        }
    }

    //todo check if needed
    //wtxn.commit();
}

// todo add transaction and cursor classes
void persipubsub::queue::Queue::init(const boost::filesystem::path& path, lmdb::env env) {
    path_ = path;

    if (env)
        env_ = std::move(env);
    else
        env_ = persipubsub::queue::initialize_environment(path, persipubsub::MAX_READER_NUM, persipubsub::MAX_DB_NUM, persipubsub::MAX_DB_SIZE_BYTES);

    {
        auto wtxn = lmdb::txn::begin(env_);

        auto pending_dbi = lmdb::dbi::open(wtxn, persipubsub::PENDING_DB, MDB_CREATE);
        auto meta_dbi = lmdb::dbi::open(wtxn, persipubsub::META_DB, MDB_CREATE);
        // lmdb::dbi::open(wtxn, persipubsub::META_DB, MDB_CREATE); // TODO ask marko how to deal with it
        auto data_dbi = lmdb::dbi::open(wtxn, persipubsub::DATA_DB, MDB_CREATE);
        auto queue_dbi = lmdb::dbi::open(wtxn, persipubsub::QUEUE_DB, MDB_CREATE);

        wtxn.commit(); // TODO this will not be needed, jumping out of scope
    }

    persipubsub::QueueData queue_data = persipubsub::lookup_queue_data(env_);

    hwm_ = HighWaterMark(queue_data.msg_timeout_secs_,
            queue_data.max_msgs_num_, queue_data.hwm_db_size_bytes_);

    strategy_ = queue_data.strategy_;

    subscriber_ids_ = std::move(queue_data.subscriber_ids_);
}


void persipubsub::queue::Queue::put(const std::string& msg, const std::vector<std::string> &subscriber_ids) const{

    vacuum();

    time_t timer;
    time(&timer);
    const auto time = boost::lexical_cast<std::string>(timer);
    boost::uuids::uuid uuid = boost::uuids::random_generator()();
    const std::string tmp = boost::uuids::to_string(uuid);

    std::string key = (time + tmp);
    lmdb::val msg_id(key.c_str());

    auto wtxn = lmdb::txn::begin(env_);

    auto pending_dbi = lmdb::dbi::open(wtxn, persipubsub::PENDING_DB);
    lmdb::val pending_subs((boost::lexical_cast<std::string>(subscriber_ids.size())).c_str());
    pending_dbi.put(wtxn, msg_id, pending_subs);

    auto meta_dbi = lmdb::dbi::open(wtxn, persipubsub::META_DB);
    lmdb::val time_val(time);
    meta_dbi.put(wtxn, msg_id, time_val);

    auto data_dbi = lmdb::dbi::open(wtxn, persipubsub::DATA_DB);
    lmdb::val data_val(msg.c_str());
    data_dbi.put(wtxn, msg_id, data_val);
    for (const auto& it: subscriber_ids) {
        auto sub_dbi = lmdb::dbi::open(wtxn, (it).c_str());
        lmdb::val sub_data("");
        sub_dbi.put(wtxn, msg_id, sub_data);
    }

    wtxn.commit();

}

void persipubsub::queue::Queue::put_many_flush_once(const std::vector<std::string> &msgs,
                         const std::vector<std::string> &subscriber_ids) const {
    vacuum();

    time_t timer;
    time(&timer);
    const auto time = boost::lexical_cast<std::string>(timer);

    auto wtxn = lmdb::txn::begin(env_);

    auto pending_dbi = lmdb::dbi::open(wtxn, persipubsub::PENDING_DB);
    auto meta_dbi = lmdb::dbi::open(wtxn, persipubsub::META_DB);
    auto data_dbi = lmdb::dbi::open(wtxn, persipubsub::DATA_DB);

    std::vector<lmdb::dbi> sub_dbis;
    for (const auto& it: subscriber_ids) {
        auto sub_dbi = lmdb::dbi::open(wtxn, (it).c_str());
        sub_dbis.push_back(std::move(sub_dbi));
    }

    for (const auto& msg: msgs) {

        boost::uuids::uuid uuid = boost::uuids::random_generator()();
        const std::string tmp = boost::uuids::to_string(uuid);
        std::string key = (time + tmp);
        lmdb::val msg_id(key.c_str());

        lmdb::val pending_subs((boost::lexical_cast<std::string>(subscriber_ids.size())).c_str());
        pending_dbi.put(wtxn, msg_id, pending_subs);
        lmdb::val time_val(time);
        meta_dbi.put(wtxn, msg_id, time_val);
        lmdb::val data_val(msg.c_str());
        data_dbi.put(wtxn, msg_id, data_val);

        for (auto& sub_dbi_it: sub_dbis) {
            lmdb::val sub_data("");
            sub_dbi_it.put(wtxn, msg_id, sub_data);
        }
    }
    wtxn.commit();

}

void persipubsub::queue::Queue::front(const std::string& identifier, std::string* msg) const{

    auto rtxn = lmdb::txn::begin(env_, nullptr, MDB_RDONLY);
    auto sub_dbi = lmdb::dbi::open(rtxn, identifier.c_str());
    auto cursor = lmdb::cursor::open(rtxn, sub_dbi);
    lmdb::val key, value;
    if (cursor.get(key, value, MDB_FIRST)) {
        auto data_dbi = lmdb::dbi::open(rtxn, persipubsub::DATA_DB);
        auto data_cursor = lmdb::cursor::open(rtxn, data_dbi);
        lmdb::val data_value;

        bool found =  data_cursor.get(key, data_value, MDB_FIRST);
        if (found){
            std::string tmp_str = data_value.data();
            std::swap(*msg, tmp_str);
        }
        else{
            cursor.close();
            rtxn.abort();
            throw std::runtime_error("Data not found");
        }
    }
    cursor.close();
    rtxn.abort();
}

void persipubsub::queue::Queue::pop(const std::string& identifier) const {
    auto wtxn = lmdb::txn::begin(env_);
    auto sub_dbi = lmdb::dbi::open(wtxn, identifier.c_str());
    auto pending_dbi = lmdb::dbi::open(wtxn, persipubsub::PENDING_DB);

    auto cursor = lmdb::cursor::open(wtxn, sub_dbi);

    lmdb::val key, value;

    if (cursor.get(key, value, MDB_FIRST)) {
        sub_dbi.del(wtxn, key);
        lmdb::val pending_value;
        pending_dbi.get(wtxn, key, pending_value);
        auto pending_num = boost::numeric_cast<unsigned int>(std::stoi(pending_value.data()));
        pending_num--;
        lmdb::val pending_new_value((boost::lexical_cast<std::string>(pending_num)).c_str());
        pending_dbi.put(wtxn, key, pending_new_value);
        cursor.close();
        wtxn.commit();
    }
    else {
        cursor.close();
        wtxn.commit();
        throw std::runtime_error("No message to pop");
    }
}


void persipubsub::queue::Queue::prune_dangling_messages() const {
    persipubsub::queue::prune_dangling_messages_for(*this, subscriber_ids_);
}


// todo size_t instead of long?
unsigned long persipubsub::queue::Queue::check_current_lmdb_size() const{
    auto rtxn = lmdb::txn::begin(env_, nullptr, MDB_RDONLY);
    auto data_dbi = lmdb::dbi::open(rtxn, persipubsub::DATA_DB);

    MDB_stat data_stat = data_dbi.stat(rtxn);

    // todo check overflow, boost::numeric_cast to unsigned long?
    unsigned long lmdb_size_bytes = data_stat.ms_psize * (data_stat.ms_branch_pages + data_stat.ms_leaf_pages + data_stat.ms_overflow_pages);

    rtxn.abort();

    return lmdb_size_bytes;

}


unsigned int persipubsub::queue::Queue::count_msgs() const {
    auto rtxn = lmdb::txn::begin(env_, nullptr, MDB_RDONLY);
    auto meta_dbi = lmdb::dbi::open(rtxn, persipubsub::META_DB);

    auto msgs_num = boost::numeric_cast<unsigned int>(meta_dbi.stat(rtxn).ms_entries);
    rtxn.abort();
    return msgs_num;
}


void persipubsub::queue::Queue::vacuum() const {
    prune_dangling_messages();

    unsigned int msgs_num = count_msgs();

    if (msgs_num >= hwm_.max_msgs_num_)
        prune_messages();
    unsigned long lmdb_size_bytes = check_current_lmdb_size();
    if (lmdb_size_bytes >= hwm_.hwm_lmdb_size_bytes_)
        prune_messages();
}


void persipubsub::queue::Queue::prune_messages() const {

    std::set<lmdb::val> messages_to_delete;

    auto rtxn = lmdb::txn::begin(env_, nullptr, MDB_RDONLY);
    auto rmeta_dbi = lmdb::dbi::open(rtxn, persipubsub::META_DB);
    auto entries = boost::numeric_cast<unsigned int>(rmeta_dbi.stat(rtxn).ms_entries);

    auto cursor = lmdb::cursor::open(rtxn, rmeta_dbi);
    lmdb::val key, value;

    if (strategy_ == persipubsub::queue::Strategy::prune_first){
        cursor.get(key, value, MDB_FIRST);
        for (unsigned int counter = 0; counter<=(entries/2); counter++){
            messages_to_delete.insert(std::move(key));
            cursor.get(key, value, MDB_NEXT);
        }
    }
    else if (strategy_ == persipubsub::queue::Strategy::prune_last){
        cursor.get(key, value, MDB_LAST);
        for (unsigned int counter = 0; counter<=(entries/2); counter++){
            messages_to_delete.insert(std::move(key));
            cursor.get(key, value, MDB_PREV);
        }
    }
    else {
        cursor.close();
        rtxn.abort();
        throw std::runtime_error("Pruning strategy not set.");
    }

    cursor.close();
    rtxn.abort();

    auto wtxn = lmdb::txn::begin(env_);
    auto pending_dbi = lmdb::dbi::open(wtxn, persipubsub::PENDING_DB);
    auto meta_dbi = lmdb::dbi::open(wtxn, persipubsub::META_DB);
    auto data_dbi = lmdb::dbi::open(wtxn, persipubsub::DATA_DB);

    std::vector<lmdb::dbi> dbis;
    dbis.push_back(std::move(pending_dbi));
    dbis.push_back(std::move(meta_dbi));
    dbis.push_back(std::move(data_dbi));

    for (const auto& sub_it: subscriber_ids_){
        auto sub_dbi = lmdb::dbi::open(wtxn, (sub_it).c_str());
        dbis.push_back(std::move(sub_dbi));
    }

    for (const auto& msg_it: messages_to_delete) {
        for (auto& db_it: dbis){
            db_it.del(wtxn, msg_it);
        }
    }

    wtxn.commit();
}

persipubsub::queue::Queue::~Queue() {}

